feat: Declarative Shadow Variables

Christopher-Chianelli opened the pull request:
VariableListeners are notoriously difficult to write, especially when you have multiple variable depending on each others.
Declarative shadow variables fixes this by providing a declarative way to define the dependencies of your shadow variables and how to calculate them.

First, annotate your variables as `@ProvidedShadowVariable`:

```java
@PlanningEntity
public class Visit {
    String id;

    @InvalidityMarker
    boolean isInvalid;   // If true, the visit is in a "loop" and some of its provided shadow variables will be null

    @ProvidedShadowVariable(MyShadowVariableProvider.class)
    LocalDateTime serviceReadyTime;
    
    // ...
}
```

Next, create a `ShadowVariableProvider` to define your variables:

```java
public class MyShadowVariableProvider implements ShadowVariableProvider {
    @Override
    public void defineVariables(ShadowVariableFactory variableFactory) {
        // The previous visit; you don't need to use @PreviousElementShadowVariable
        var previousVisit = variableFactory.entity(Visit.class).previous();
        // The vehicle the visit is assignedf to; you don't need to use @InverseRelationShadowVariable
        var vehicle = variableFactory.entity(Visit.class).inverse(Vehicle.class);
        
        // The service ready time; when the technician will arrive
        var serviceReadyTime = variableFactory.newShadow(Visit.class)
               // If a visit has a previous visit, the ready time 
                .compute(previousVisit.fact(Location.class, Visit::getLocation),
                         previousVisit.variable(LocalDateTime.class, "serviceFinishTime"),
                         (visit, previousLocation, previousEndTime) -> previousEndTime.plus(visit.getLocation().travelTimeFrom(previousLocation))
                )
                // If a visit has no previous visit, use the vehicle's start time
                .orCompute(vehicle.fact(Location.class, Vehicle::getLocation),
                        vehicle.fact(LocalDateTime.class, Vehicle::getStartTime),
                        (visit, startLocation, startTime) -> startTime.plus(visit.getLocation().travelTimeFrom(startLocation))
                )
                // If a visit is unassigned, its serviceReadyTime is null
                .as("serviceReadyTime");
        
         // The service start time; when service can start
         // If the visit is a part of a group, service cannot start until all technicians arrive
         // Otherwise, it the same as the service start time
         var visitGroup = variableFactory
                        .entity(Visit.class)
                        .group(Visit.class, Visit::getVisitGroup);
         var serviceStartTime = variableFactory.newShadow(Visit.class)
                // If the visit has a visit group, its start time is the max of the ready times of the visit in its visit group
                .compute(visitGroup.variables(LocalDateTime.class, "serviceReadyTime"),
                        (visit, groupReadyTimes) -> groupReadyTimes.isEmpty() ? null : Collections.max(groupReadyTimes))
                // Otherwise, it the service ready time of this visit
                .orCompute(serviceReadyTime, (visit, readyTime) -> readyTime)
                .as("serviceStartTime");

         // Each service ends 30 minutes after it started
         var serviceFinishTime = variableFactory.newShadow(Visit.class)
                .compute(serviceStartTime, (visit, startTime) -> startTime.plusMinutes(30))
                .as("serviceFinishTime");
    }
}
```

and finally, write tests to ensure your shadow variables work correctly:

```java
@Test
public void shadowVariables() {
      var sessionFactory = ShadowVariableSessionFactory.create(
              SolutionDescriptor.buildSolutionDescriptor(RoutePlan.class,
                      Vehicle.class, Visit.class),
              new MyShadowVariableProvider());

      var vehicle = Vehicle("v1");
      var visit1 = Visit("c1");
      var visit2 = Visit("c2");
      var visit3 = Visit("c3");

      var session = sessionFactory.forEntities(vehicle, visit1, visit2, visit3);
      session.setInverse(visit1, vehicle);
      session.setPrevious(visit2, visit1);
      session.setPrevious(visit3, visit2);
      session.updateVariables();

      assertThat(visit1.getServiceReadyTime()).isEqualTo(TestShadowVariableProvider.BASE_START_TIME);
      assertThat(visit1.getServiceStartTime()).isEqualTo(TestShadowVariableProvider.BASE_START_TIME);
      assertThat(visit1.getServiceFinishTime()).isEqualTo(TestShadowVariableProvider.BASE_START_TIME.plusMinutes(30L));
      assertThat(visit1.isInvalid()).isFalse();

      assertThat(visit2.getServiceReadyTime()).isEqualTo(TestShadowVariableProvider.BASE_START_TIME.plusMinutes(60L));
      assertThat(visit2.getServiceStartTime()).isEqualTo(TestShadowVariableProvider.BASE_START_TIME.plusMinutes(60L));
      assertThat(visit2.getServiceFinishTime()).isEqualTo(TestShadowVariableProvider.BASE_START_TIME.plusMinutes(90L));
      assertThat(visit2.isInvalid()).isFalse();

      assertThat(visit3.getServiceReadyTime()).isEqualTo(TestShadowVariableProvider.BASE_START_TIME.plusMinutes(120L));
      assertThat(visit3.getServiceStartTime()).isEqualTo(TestShadowVariableProvider.BASE_START_TIME.plusMinutes(120L));
      assertThat(visit3.getServiceFinishTime()).isEqualTo(TestShadowVariableProvider.BASE_START_TIME.plusMinutes(150L));
      assertThat(visit3.isInvalid()).isFalse();
}
```

Behind the scenes, Timefold Solver calculates a valid topological order for each of your shadow variables.
This allows Timefold Solver to:

1. Only recalculate changed shadow variables. A shadow variable is recalculated if any of its declared inputs has changed. Using an undeclared variable input will lead to score corruptions.
2. Determine what variables are in a "loop". Variables are in a "loop" if:

- They directly or indirectly depend on each other.
- They depend on looped variables.

Currently a WIP:

- [ ] Write javadoc
- [ ] Write documentation
- [ ] Write additional tests
- [ ] Write additional use cases
- [ ] Finalize API

Christopher-Chianelli committed:
chore: create initial graph backend for provided shadow variables

Provided shadow variables work by calculating the topological order
of each shadow variable.

The nodes in the graph are paths to each shadow variable, binded to a
particular entity instance.

- The path `e1:Entity.#id.a` is the source path for the shadow variable `a` on entity e1
- If `e2.previous = e1`, then the path `e2:Entity.#id.#previous.a` is an alias path for the shadow
  variable `a` on e1.
- The path can have multiple parts; like
  `e1:Entity.#id.#previous.#previous.a`. In this case,
  `e1:Entity.#id.#previous` is the parent of
  `e1:Entity.#id.#previous.#previous`.

The edges in the graph are the dependencies for each shadow variable.

- There is a fixed edge from the parent to each of its children.
  (i.e. `e1:Entity.#id.#previous` -> `e1:Entity.#id.#previous.a`)
- There is a fixed edge from the direct dependencies of a shadow variable to the shadow variable.
  (i.e. `e1:Entity.#id.#previous.readyTime` -> `e1:Entity.#id.#startTime`)
- There is a dynamic edge from each shadow variable to all its aliases.
  (i.e. `e1:Entity.#id.startTime` ->
  `e2:Entity.#id.#previous.startTime`, if e1 is the previous of e2.)

Tarjan's algorithm is used to calculate the topological order of each
node. Once the topological order of each node is known, to update from
a set of changes:

1. Pick a changed node with the minimum topological order that was not
   visited.
2. Update the changed node.
3. If the value of the node changed, marked all its children as changed.

Christopher-Chianelli committed:
chore: add support for grouping

- Only facts of entities can be group
- If e is in a group on g, then there is an edge from `e:Entity.#id.shadow` to `g:Entity.#id.group.shadow`.

Christopher-Chianelli committed:
chore: start integrating the graph into the score director

Christopher-Chianelli committed:
chore: fix assertion methods, mark everything as changed

ESC is consistent with Assertion ESC for visit group if we
recalculate everything from scratch, which implies the graph
edges/topological order is correct, but not everything that is
changed is marked as changed.

Christopher-Chianelli committed:
chore: Mark unassigned elements as changed + ignore unassigned visits in ESC

VisitGroups are now working on FULL_ASSERT

Christopher-Chianelli committed:
chore: Rename files, add some docs, add VariableListenerSupport test

Christopher-Chianelli committed:
chore: rename API method to be more accurate

Christopher-Chianelli committed:
chore: Fix Javadoc element

triceo commented:
Before I do a proper review, I suggest we do some house-keeping first. (Otherwise comments would get lost when files are being renamed and moved around.)

- Are we happy with the package? Variables generally go into `api.domain.variable`, so in this case `preview.api.domain.variable.declarative`?
- Preview features need to be declared in `PreviewFeature`; features not enabled through this enum need to fail fast when used. (In this case, probably when the new annotations are being processed.)

triceo commented on core/src/main/java/ai/timefold/solver/core/impl/domain/entity/descriptor/EntityDescriptor.java:
```diff
@@ -637,10 +663,21 @@ public boolean hasVariableDescriptor(String variableName) {
+        var variableDescriptor = effectiveVariableDescriptorMap.get(variableName);
+        if (variableDescriptor == null) {
+            throw new IllegalArgumentException(
+                    "Entity class %s does not hava a \"%s\" genuine or shadow variable. Maybe you meant one of %s?"
```
Using raw strings (`"""`) will let you remove the escaping.
Also, the maybe should be on a new line, which raw strings also make easier.

triceo commented on core/src/main/java/ai/timefold/solver/core/impl/domain/variable/provided/AbstractVariableReference.java:
```diff
@@ -0,0 +1,23 @@
+import org.jspecify.annotations.NonNull;
+import org.jspecify.annotations.Nullable;
+
+public abstract sealed class AbstractVariableReference<Solution_, Entity_, Value_> implements VariableReference<Entity_, Value_>
```
It says class, but it doesn't actually have any concrete methods, which suggests it's an interface instead. :-) We typically call these interfaces `InnerX`, where `X` is the public thing that's being extended.

triceo commented on core/src/main/java/ai/timefold/solver/core/impl/domain/variable/provided/ChangedVariableNotifier.java:
```diff
@@ -0,0 +1,39 @@
+        this.afterVariableChanged = afterVariableChanged;
+    }
+
+    public static <Solution_> ChangedVariableNotifier<Solution_> empty() {
```
Arguably, this can be a singleton.

triceo commented on core/src/main/java/ai/timefold/solver/core/impl/domain/variable/provided/ChangedVariableNotifier.java:
```diff
@@ -0,0 +1,39 @@
+                });
+    }
+
+    public static <Solution_> ChangedVariableNotifier<Solution_> of(InnerScoreDirector<Solution_, ?> scoreDirector) {
```
And this can be stored on a score director, meaning there would be a 1:1 mapping between these instances and score directors.

triceo commented on core/src/main/java/ai/timefold/solver/core/impl/domain/variable/provided/DefaultGroupVariableReference.java:
```diff
@@ -0,0 +1,244 @@
+public final class DefaultGroupVariableReference<Solution_, Entity_, ParentValue_, Value_>
+        extends AbstractVariableReference<Solution_, Entity_, List<Value_>>
+        implements GroupVariableReference<Entity_, Value_> {
+    @NonNull
```
Let's start using a class-level `@NullMarked` - that eliminates the need to use `@NonNull`.

triceo commented on core/src/main/java/ai/timefold/solver/core/impl/domain/variable/provided/DefaultShadowCalculationBuilder.java:
```diff
@@ -0,0 +1,114 @@
+import ai.timefold.solver.core.preview.api.variable.provided.SingleVariableReference;
+import ai.timefold.solver.core.preview.api.variable.provided.VariableReference;
+
+public class DefaultShadowCalculationBuilder<Solution_, Entity_, Value_> implements ShadowCalculationBuilder<Entity_, Value_> {
```
Also, you seem to be using the nullability markers somewhat selectively.
Why are they used in one place, and not in other places?

Christopher-Chianelli replied:
Mostly, they are used in classes where nullability is mixed (i.e. some fields/methods can be null, others cannot).

triceo commented on core/src/main/java/ai/timefold/solver/core/impl/domain/variable/provided/InvalidityMarkerVariableDescriptor.java:
```diff
@@ -0,0 +1,52 @@
+
+    @Override
+    public List<VariableDescriptor<Solution_>> getSourceVariableDescriptorList() {
+        return List.of();
```
IMO `Collections.emptyList()` better describes what will happen here.

triceo commented on core/src/main/java/ai/timefold/solver/core/preview/api/variable/provided/GroupVariableReference.java:
```diff
@@ -0,0 +1,19 @@
+import java.util.List;
+import java.util.function.Function;
+
+public interface GroupVariableReference<Entity_, Value_> extends VariableReference<Entity_, List<Value_>> {
```
Arguably, the public APIs need the nullness markers the most.

triceo commented on core/src/main/java/ai/timefold/solver/core/preview/api/variable/provided/ShadowVariableSession.java:
```diff
@@ -0,0 +1,13 @@
+package ai.timefold.solver.core.preview.api.variable.provided;
+
+public interface ShadowVariableSession {
```
My understanding of this API is still limited, but do we actually need this public API anywhere in the solver? IMO this is internal in core, and only exposed in the tests; therefore I'd make this public in the tests, not in the solver itself.

Christopher-Chianelli replied:
It can be moved; it is here for convenience when testing the prototype. (core internally depends on the default implementation which has before/after methods).

triceo commented on core/src/main/java/ai/timefold/solver/core/preview/api/variable/provided/ShadowVariableSessionFactory.java:
```diff
@@ -0,0 +1,13 @@
+import ai.timefold.solver.core.impl.domain.solution.descriptor.SolutionDescriptor;
+import ai.timefold.solver.core.impl.domain.variable.provided.MockShadowVariableSessionFactory;
+
+public interface ShadowVariableSessionFactory {
```
Dtto.

triceo commented:
Hypothetical scenario:

- I have two different shadow providers.
- But they end up declaring/depending on the same shadow variables.

Do they all become part of the same graph? In other words - is the graph limited to one shadow provider, or is it built from all shadow providers combined, and then split into individual independent components?

Christopher-Chianelli commented:
> Hypothetical scenario:
> 
>     * I have two different shadow providers.
> 
>     * But they end up declaring/depending on the same shadow variables.
> 
> 
> Do they all become part of the same graph? In other words - is the graph limited to one shadow provider, or is it built from all shadow providers combined, and then split into individual independent components?

Although untested, it is built from all shadow providers combined, and then split into individual independent components. Defining the same shadow variable twice is an error though; (you can use it multiple time, but cannot declare it multiple times).

triceo commented:
Please add this to the list of things that need to be tested.

Christopher-Chianelli committed:
chore: Use @NullMarked, move package, check if preview feature enabled

Christopher-Chianelli committed:
fix: only override the default preview feature set if it is non-null

Christopher-Chianelli committed:
chore: Rename methods, allow null values, fix multiple fallbacks

ge0ffrey commented on core/src/main/java/ai/timefold/solver/core/preview/api/domain/variable/declarative/InvalidityMarker.java:
```diff
@@ -0,0 +1,13 @@
+
+@Target({ METHOD, FIELD })
+@Retention(RUNTIME)
+public @interface InvalidityMarker {
```
Will this mark any type of invalidity or only loops?

triceo replied:
TODO: Naming and Javadoc.

ge0ffrey commented on core/src/main/java/ai/timefold/solver/core/preview/api/domain/variable/declarative/ShadowVariableSession.java:
```diff
@@ -0,0 +1,17 @@
+import org.jspecify.annotations.Nullable;
+
+@NullMarked
+public interface ShadowVariableSession {
```
This seems similar to ConstraintVerifier, so the naming should be similar.
Does the ConstraintVerifier API have anything that is a Session?

triceo replied:
As far as I understand, this does not need to be in the public API. Please remove.

ge0ffrey commented on core/src/main/java/ai/timefold/solver/core/preview/api/domain/variable/declarative/ShadowVariableSessionFactory.java:
```diff
@@ -0,0 +1,16 @@
+        return new MockShadowVariableSessionFactory<>(solutionDescriptor, shadowVariableProvider);
+    }
+
+    ShadowVariableSession forEntities(Object... entities);
```
Feels similar to ConstraintVerifier.given(), so maybe rename it to "given"?

ge0ffrey commented on core/src/main/java/ai/timefold/solver/core/preview/api/domain/variable/declarative/ShadowVariableSessionFactory.java:
```diff
@@ -0,0 +1,16 @@
+import org.jspecify.annotations.NullMarked;
+
+@NullMarked
+public interface ShadowVariableSessionFactory {
```
Maybe rename to something like ShadowVariableVerifier, as it serves a a simular purpose?

triceo replied:
Likewise, please remove from public API. The user has no use for this.

ge0ffrey commented on core/src/test/java/ai/timefold/solver/core/impl/testdata/domain/declarative/fsr/TestdataFSRAssertionConstraintProvider.java:
```diff
@@ -0,0 +1,30 @@
+
+import org.jspecify.annotations.NonNull;
+
+public class TestdataFSRAssertionConstraintProvider implements ConstraintProvider {
```
Is FSR being copy pasted into the solver test sources?
Or is it a subset?

Christopher-Chianelli replied:
A **heavily** simplified subset.

triceo replied:
Please remove references to FSR. FSR has no meaning in the context of the solver.

Christopher-Chianelli committed:
chore: Remove workaround for bug in undo moves

Christopher-Chianelli committed:
feat: Add support for intermediate shadow variables

Christopher-Chianelli committed:
fix: Handle group items being a different entity than the parent entity

Christopher-Chianelli committed:
fix: merge conflicts

Christopher-Chianelli committed:
fix: Null handling, Quarkus preview features, method variable descriptors

Christopher-Chianelli committed:
chore: Move to the new API

- Added fail-fasts for prohibited paths (that probably also did not work
  on the old impl anyway)

Christopher-Chianelli committed:
chore: merge fixes

triceo commented on core/src/main/java/ai/timefold/solver/core/impl/domain/variable/listener/support/VariableListenerSupport.java:
```diff
@@ -111,6 +127,7 @@ private void processShadowVariableDescriptorWithListVariable(ShadowVariableDescr
 
     private void
             processShadowVariableDescriptorWithoutListVariable(ShadowVariableDescriptor<Solution_> shadowVariableDescriptor) {
+        // TODO: Set up variable listener for declarative shadow variables
```
Should this still be here? I'd expect this feature to be in a mergable state, and therefore without TODOs. Possible future improvements will be tracked separately.

triceo commented on core/src/main/java/ai/timefold/solver/core/preview/api/domain/variable/declarative/ShadowVariableUpdater.java:
```diff
@@ -0,0 +1,12 @@
+
+@Target(ElementType.METHOD)
+@Retention(RetentionPolicy.RUNTIME)
+public @interface ShadowVariableUpdater {
```
TODO: Naming and Javadoc

triceo commented on core/src/test/java/ai/timefold/solver/core/impl/testdata/domain/declarative/task_assignment/TestdataTAConstraintProvider.java:
```diff
@@ -0,0 +1,27 @@
+
+import org.jspecify.annotations.NonNull;
+
+public class TestdataTAConstraintProvider implements ConstraintProvider {
```
Just as with "FSR", "TA" is also meaningless here. The solver isn't tailored to any particular problem, and these domain classes should therefore carry generic names. We already have dozens of them, and we were able to follow this policy; I don't see the need to divert now.

triceo reviewed (commented):
First round of review, surface-level.
Next round will look deeper into the logic.

triceo commented on core/src/main/java/ai/timefold/solver/core/impl/domain/entity/descriptor/EntityDescriptor.java:
```diff
@@ -352,7 +366,14 @@ The entityClass (%s) has a @%s annotated member (%s) that has an unsupported typ
+            ShadowVariableDescriptor<Solution_> variableDescriptor;
+            var annotation = memberAccessor.getAnnotation(ShadowVariable.class);
+            if (annotation != null && !annotation.method().isEmpty()) {
+                variableDescriptor =
```
Where are the fail-fasts for when the other fields are used? The Javadoc suggests that they exist, but I didn't notice them.

Christopher-Chianelli replied:
Fail fasts are in RootVariableSource, which does the processing of the variable paths.

triceo commented on core/src/main/java/ai/timefold/solver/core/impl/domain/variable/declarative/DefaultShadowVariableSession.java:
```diff
@@ -0,0 +1,70 @@
+public class DefaultShadowVariableSession<Solution_> implements ShadowVariableSession, Supply {
+    final VariableReferenceGraph<Solution_> graph;
+
+    record EntityVariablePair<Solution_>(VariableDescriptor<Solution_> variableDescriptor, Object entity) {
```
Considering what the descriptor is used for, it can easily be replaced by VariableMetaModel.

triceo commented on core/src/main/java/ai/timefold/solver/core/impl/domain/variable/declarative/DefaultShadowVariableSessionFactory.java:
"Accessing a declarative variable from another declarative variable" clearly doesn't mean what I think it means, because accessing one variable for another is the goal of this entire "a.b" syntax.

Christopher-Chianelli replied:
The older version the PR was not tested with more exotic cases, such as "previous.previous" or "factProperty.previous" or "myShadowVariable.myOtherShadowVariable".

The old impl made it harder to see the inherent issues that this impl makes obvious by removing all the indirection.

Basically, the three rules are:

- You can only access normal shadow variables from the root entity or from a collection element. Otherwise, the variable listener need to maintain an inverse set to lookup the affected entities for a shadow variable change.
- You cannot chain multiple declarative shadow variables together in a single path. For instance, "closestStation.visitCount"; this has the potential to change the dependency/topology of the graph in the middle of an update.
- Your path must have at least one variable in it, and all names must be the name of members in the class.

triceo replied:
Two comments:

- Let's make sure these rules are listed in a Javadoc somewhere. Lately we've been touching decade-old code, and we often wished the authors of that code had the foresight to list all of their thinking, rules and assumptions somewhere. Code archeology is not fun.

- "Your path must have at least one variable in it, and all names must be the name of members in the class." What else _could_ be there in the path? I'd argue _nothing but variables_. Constants do not need to be referenced - we only need to reference what we need to build and maintain the graph.

triceo commented on core/src/main/java/ai/timefold/solver/core/impl/domain/variable/declarative/DefaultShadowVariableSessionFactory.java:
```diff
@@ -0,0 +1,185 @@
+        for (var declarativeShadowVariable : declarativeShadowVariableDescriptors) {
+            var fromEntityClass = declarativeShadowVariable.getEntityDescriptor().getEntityClass();
+            var fromVariableName = declarativeShadowVariable.getVariableName();
+            final var fromVariableId = new VariableId(fromEntityClass, fromVariableName);
```
VariableMetaModel was invented earlier to serve this purpose.
(VariableDescriptor is more of a utility class, and it is a mess. VariableMetaModel is a well-defined carrier of metadata.)

triceo commented on core/src/main/java/ai/timefold/solver/core/impl/domain/variable/declarative/DefaultShadowVariableSessionFactory.java:
```diff
@@ -0,0 +1,185 @@
+            }
+        }
+
+        // Create variable listeners for each declarative shadow variable descriptor
```
Where exactly are the listeners being created? I don't see any use of `VariableListener` here. Maybe let's call it something else, to not mix concepts?

triceo commented on core/src/main/java/ai/timefold/solver/core/impl/domain/variable/declarative/DefaultShadowVariableSessionFactory.java:
This is a strange cast. I don't think I've ever seen a double cast, honestly didn't know it was even possible. What's the point?

triceo commented on core/src/main/java/ai/timefold/solver/core/impl/domain/variable/declarative/DefaultTopologicalOrderGraph.java:
```diff
@@ -0,0 +1,152 @@
+    @Override
+    public void endBatchChange() {
+        AtomicInteger index = new AtomicInteger(1);
+        AtomicInteger stackIndex = new AtomicInteger(0);
```
I think I introduced a `MutableInteger` a while back. Since you don't need atomicity, you can use that one.

triceo commented on core/src/main/java/ai/timefold/solver/core/impl/domain/variable/declarative/DefaultTopologicalOrderGraph.java:
```diff
@@ -0,0 +1,152 @@
+            int current;
+            do {
+                current = stack[stackIndex.decrementAndGet()];
+                onStackSet[current] = false;
```
I don't understand - if `out` is a `BitSet`, why is `onStackSet` (which also stores bits) not a `BitSet` as well?

Christopher-Chianelli replied:
onStackSet is simply a mapping from node (an int)  and does not need any of the BitSet methods. Essentially, indexMap, lowMap, and onStackSet acts as additional attributes on nodes.

out is a BitSet since we need to iterate it and access other BitSet methods.

triceo commented on core/src/main/java/ai/timefold/solver/core/impl/domain/variable/declarative/EntityVariableOrFactReference.java:
```diff
@@ -0,0 +1,23 @@
+
+import java.util.Objects;
+
+public record EntityVariableOrFactReference<Entity_>(VariableId variableId, Entity_ entity,
```
Another possible use of VariableMetaModel.

triceo commented on core/src/main/java/ai/timefold/solver/core/impl/domain/variable/declarative/RootVariableSource.java:
```diff
@@ -0,0 +1,246 @@
+            String variablePath,
+            MemberAccessorFactory memberAccessorFactory,
+            DescriptorPolicy descriptorPolicy) {
+        var pathParts = variablePath.split("\\.");
```
If we have `COLLECTION_REFERENCE_SUFFIX`, let's make the separator a constant too.

triceo commented on core/src/main/java/ai/timefold/solver/core/impl/domain/variable/declarative/RootVariableSource.java:
```diff
@@ -0,0 +1,246 @@
+                    throw new IllegalArgumentException("Cannot reference a collection on a variable.");
+                }
+                if (hasListMemberAccessor) {
+                    throw new IllegalArgumentException("Cannot reference a collection on a collection.");
```
Deserves a more descriptive message.
Where did the user make a mistake? Maybe there is a suggestion for fixing it?

triceo replied:
Also applies to the previous exception immediately above.

triceo commented on core/src/main/java/ai/timefold/solver/core/impl/domain/variable/declarative/RootVariableSource.java:
```diff
@@ -0,0 +1,246 @@
+                    downstreamDeclarativeVariable,
+                    sourceVariablePath.getValueVisitorFromVariableEntity()));
+            isTopLevel = false;
+        }
```
Arguably can be extracted from this already very long method.

triceo commented on core/src/main/java/ai/timefold/solver/core/impl/domain/variable/declarative/RootVariableSource.java:
```diff
@@ -0,0 +1,246 @@
+                    consumer.accept((Value_) current);
+                }
+            };
+        }
```
Arguably can be extracted from this already very long method.

triceo commented on core/src/main/java/ai/timefold/solver/core/impl/domain/variable/declarative/RootVariableSource.java:
```diff
@@ -0,0 +1,246 @@
+                }
+
+                Iterable<Object> iterable = (Iterable<Object>) current;
+                outer: for (var item : iterable) {
```
Can we do this without labels? My experience has been that, whenever I felt like I need to use explicit goto, there was a better code structure that avoided it.

triceo commented on core/src/main/java/ai/timefold/solver/core/impl/domain/variable/declarative/VariableReferenceGraph.java:
```diff
@@ -0,0 +1,305 @@
+            // Variable state is probably incorrect/will trigger a different
+            // exception, so return early.
+            return;
+        }
```
If an uncaught exception occurred, why are we even here? The solver would've crashed before getting here. (The undo move issue was already fixed.)

triceo commented on core/src/main/java/ai/timefold/solver/core/impl/domain/variable/declarative/VariableReferenceGraph.java:
```diff
@@ -0,0 +1,305 @@
+            return;
+        }
+
+        record AffectedEntity(Object entity, VariableUpdaterInfo variableUpdaterInfo) {
```
The method is long enough already, even without data types being declared inside of it. Please move this type outside - you can still make it private.

triceo commented on core/src/test/java/ai/timefold/solver/core/preview/api/variable/declarative/fsr/FSRShadowVariableTest.java:
```diff
@@ -0,0 +1,491 @@
+import org.junit.jupiter.api.Test;
+import org.mockito.Mockito;
+
+public class FSRShadowVariableTest {
```
These tests are complex enough that they deserve comments.
What behavior is being tested by each group of lines, why is the expected result what it is.

triceo commented on core/src/test/java/ai/timefold/solver/core/preview/api/variable/declarative/task_assignment/TAShadowVariableTest.java:
```diff
@@ -0,0 +1,72 @@
+
+import org.junit.jupiter.api.Test;
+
+public class TAShadowVariableTest {
```
Let's not overdo it with the tests that run the solver.
I'm fine with having this one, it's a good end-to-end test, but let's test future functionality on a much more granular level.

triceo reviewed (changes requested):
General comments having read the entire PR in depth:

- The code structure could be improved. There are several very long methods, which (when broken down into smaller, well-named ones) will help the algorithms to be self-documenting.
- The public API of the feature is undocumented.
- The testing infrastructure for the new mechanism should be hidden for now; we haven't discussed it at all, and we should not expose it to the users since we only have a limited amount of time to get this merged.
- The tests are complex enough to warrant comments.

We need to go through a naming excercise for the two new annotations we expose. Also, the feature name itself ("declarative shadow variables") is arguably no longer accurate. Let's leave that for the end, once everything is properly documented, at which point we'll have the most accurate information to infer correct names for concepts.

Christopher-Chianelli committed:
perf: use VariableDescriptor.ordinal instead of VariableDescriptor.hashCode

For large datasets, EntityVariablePair.hashCode dominated the
flamegraph.

triceo commented on core/src/main/java/ai/timefold/solver/core/impl/domain/variable/declarative/DefaultShadowVariableSession.java:
```diff
@@ -19,14 +18,14 @@ record EntityVariablePair<Solution_>(VariableDescriptor<Solution_> variableDescr
         @Override
         public int hashCode() {
-            return Objects.hash(System.identityHashCode(entity), variableDescriptor);
+            return (31 * System.identityHashCode(entity)) ^ variableDescriptor.getOrdinal();
```
... but ordinal has the function. In either case, arguably we should put this logic inside `VariableDescriptor` and not externalize it here.

triceo replied:
And maybe I would add a comment here that this class is often used as key in hashmaps, and therefore this method needs to be optimized.

Christopher-Chianelli commented on core/src/main/java/ai/timefold/solver/core/impl/domain/variable/declarative/RootVariableSource.java:
```diff
@@ -0,0 +1,246 @@
+        }
+
+        if (variableSourceReferences.isEmpty()) {
+            throw new IllegalArgumentException("The source path \"%s\" on entity class %s does not reference any variables."
```
Even for arguments that are Strings?  (i.e. should Strings be represented as `"text"`, `(text)` or `("text")`?

triceo replied:
I don't think we add the quotes for strings. At least I don't recall a situation where we do. In fact, most things we put here are some sort of a string (class name, var name, ...) and we do not use quotation marks for that.

triceo commented on core/src/main/java/ai/timefold/solver/core/impl/domain/variable/declarative/VariableReferenceGraph.java:
```diff
@@ -0,0 +1,305 @@
+        changed.set(node.id());
+    }
+
+    public void updateChanged() {
```
This method is very long and also - not coincidentally - shows up as a dominator in the flamegraph. Please break this method down into smaller methods, so that we can better understand the logic, and so that the flame graph can actually meaningfully show how much each part contributes to the result.

Christopher-Chianelli committed:
chore: address some review comments

Christopher-Chianelli committed:
chore: use VariableMetamodel instead of VariableId

Christopher-Chianelli committed:
chore: rename FSR/TA tests/classes, add comments to FSR tests, javadoc for TopologicalOrderGraph.

Christopher-Chianelli committed:
chore: fix Sonar issues + renaming

triceo commented on core/src/test/java/ai/timefold/solver/core/impl/testdata/domain/declarative/concurrent_values/TestdataConcurrentValue.java:
```diff
@@ -0,0 +1,272 @@
+        return startTime;
+    }
+
+    @ShadowSources(sources = { "serviceStartTime" })
```
Let's try to do it without the mandatory argument name. With just one argument, this should be possible.

Christopher-Chianelli committed:
docs: Add Javadoc for annotations

Christopher-Chianelli commented on benchmark/src/main/resources/benchmark.xsd:
```diff
@@ -2603,6 +2603,9 @@
     <xs:restriction base="xs:string">
                   
       
+      <xs:enumeration value="DECLARATIVE_SHADOW_VARIABLES"/>
```
What about "supplier variable"? Matches the API syntax `@ShadowVariable(supplierName="...")`?

triceo replied:
I was thinking a bit more, and let's just keep the name for now.
While it's in preview, it's going to be called this.
Once it's no longer in preview, it's just a shadow variable, and all other options are deprecated. Even the annotation says "ShadowVariable", nothing else.
So the need to provide another name will go away together with the preview status.

triceo commented on core/src/main/java/ai/timefold/solver/core/impl/domain/variable/declarative/VariableReferenceGraph.java:
```diff
@@ -0,0 +1,330 @@
+
+            if (isChanged) {
+                graph.nodeForwardEdges(nextNode).forEachRemaining(
+                        (int node) -> nodeHeap.add(new AffectedShadowVariable(node, graph.getTopologicalOrder(node))));
```
Performance idea as I was reading through this: could it be possible that a node is being added even though it is already `visited`? If so, let's not add it, save ourselves some iteration.

triceo commented on core/src/main/java/ai/timefold/solver/core/impl/domain/variable/declarative/VariableReferenceGraph.java:
```diff
@@ -0,0 +1,330 @@
+    @SuppressWarnings("unchecked")
+    private void updateInvalidityStatusOfAffectedEntities(Set<AffectedEntity> affectedEntities, LoopedTracker loopedTracker) {
+        for (var affectedEntity : affectedEntities) {
+            var invalidDescriptor = affectedEntity.variableUpdaterInfo.invalidityMarkerVariableDescriptor();
```
Please look for the old name "invalidityMarker" and replace it with the new name.

triceo commented on core/src/main/java/ai/timefold/solver/core/impl/domain/variable/declarative/VariableReferenceGraph.java:
```diff
@@ -0,0 +1,330 @@
+            var topologicalOrder = graph.getTopologicalOrder(i);
+            heap.add(new AffectedShadowVariable(i, topologicalOrder));
+            if (i == Integer.MAX_VALUE) {
+                break; // or (i+1) would overflow
```
Shouldn't we rather throw an exception here?
Seems like this is a failure state, maybe even impossible state.

Christopher-Chianelli replied:
Overly cautious `BitSet` iteration. How `BitSet` iteration works is by querying for the first set bit after or at an index. Then to get the next set bit, you do `nextSetBit(currentBit + 1)`. `currentBit + 1` will overflow if BitSet has the INT_MAX bit set. For our cases, this is impossible, since the JVM limit array sizes to slightly less than INT_MAX.

triceo replied:
I think you've just given a reason for why this should have been treated as a failure state. :-) At the very least, there should have been a comment explaining what you just described.

Try to think of it like this: if I have a question because I don't understand something, it's probably a piece of code that deserves an explanatory comment. :-) There already is a comment, but in this case, the additional context you provided is useful.

triceo commented on core/src/main/java/ai/timefold/solver/core/impl/domain/variable/declarative/VariableReferenceGraph.java:
```diff
@@ -0,0 +1,330 @@
+            Entity_ entity,
+            VariableUpdaterInfo variableReference) {
+        if (!variableId.entity().type().isInstance(entity)) {
+            throw new IllegalArgumentException(variableId + " " + entity);
```
Deserves a descriptive message.

Christopher-Chianelli replied:
Leftover from debugging; removing

triceo reviewed (approved):
LGTM after all conversations are resolved. 
(I added a small amount of new ones.)

triceo commented on core/src/main/java/ai/timefold/solver/core/preview/api/domain/variable/declarative/ShadowSources.java:
```diff
@@ -0,0 +1,98 @@
+     * "variableName", for referring any variable on the same planning entity.
+     * </li>
+     * <li>
+     * A list of names seperated by ".", such as "variableOrFact.fact.entity.supplierVariable",
```
Why would you ever reference a fact? This sounds like something that should not be possible.

Christopher-Chianelli replied:
FSR needs this due to the domain; the visit collection of a visit group is on its own object instead of on the entity; i.e. `visitGroup.visits[]` instead of `visitGroup[]`.

triceo replied:
Then arguably let's only allow that use case, and forbid all other fact references?

You don't need to reference `factA.factB` for correctness - you only ever need to reference `factA.facts[]`. A chain of fact references is IMO unnecessary and I'd rather forbid it now, possibly re-introducing it later if there's demand. 

IMO we should not encourage code with so many levels of indirection.

Christopher-Chianelli replied:
It is arguably more code to allow only a single level of indirection/fail otherwise and for very little benefit. IMO if we support one level of indirection, we might as well support infinite levels of indirection, since it takes no extra code to support (and it takes extra code to explicitly forbid).

Christopher-Chianelli replied:
Java encourages indirection, so we _should_ allow it.
It is extremely reasonable to, in Java, have code structure like this:

```java
public record ConcurrentTasks(List<Task> tasks, TimeAdjuster adjuster) {}
public record DependencyTasks(List<Task> tasks, TimeAdjuster adjuster) {}
public record TaskGroup(ConcurrentTasks concurrentTasks, DependencyTasks dependencyTasks) {}

@PlanningEntity
public class Task {
    TaskGroup taskGroup;
}
```

Such a structure would require at least two level of indirections. Last I checked, we wanted users to be able to use their own classes without restrictions. Imposing a "at most one level of indirection in planning entity classes" is an unreasonable restriction that will force users to create their own, separate Solver specific classes.

triceo replied:
We limit users every day in what sort of models they can build. We have rules about inheritance, and what methods you can and can not override. We do that to keep complexity at hand.

We also make decisions that limit users in other places. We only have quad streams, assuming no more will ever be necessary. (And for the one person who ever asked for more, we created `map(...)`.

Is one level of indirection too low a limit? Maybe. I'd rather go too low and have to increase the limit later, than go too high and enable unnecessary complexity. 

Arguably, this is a philosophical debate - are we a project that allows users unlimited power, and bets on their responsibility and wisdom? Or are we a project that has an opinion, and makes conscious choices which guide people towards certain preferred outcomes? In my opinion, we have always been the former.

Christopher-Chianelli replied:
If we are the latter, we are doing a terrible job at it; we are effectively the former with SolverConfig + what we allow in user models.

The SolverConfig gives the user infinite freedom, is vital to provide/edit in order to get the best performance, and does not provide an opinion on anything.

We do not forbid the vast majority of user code; we only forbid user code that is incompatible with the Solver (ex: code w/o a no-arg constructor or code referring to a variable that does not exist).

Christopher-Chianelli committed:
chore: review comments, add tests for RootVariableSource

Christopher-Chianelli committed:
chore: review comments
